Userland/Test.cpp

```diff
@@ -0,0 +1,540 @@
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <AK/NonnullOwnPtr.h>
+#include <AK/OwnPtr.h>
+#include <AK/Vector.h>
+#include <LibCore/DirIterator.h>
+#include <LibCore/File.h>
+#include <getopt.h>
+#include <grp.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <unistd.h>
+
+bool g_there_was_an_error = false;
+
+[[noreturn]] void fatal_error(const char* format, ...)
+{
+    fputs("\033[31m", stderr);
+
+    va_list ap;
+    va_start(ap, format);
+    vfprintf(stderr, format, ap);
+    va_end(ap);
+
+    fputs("\033[0m\n", stderr);
+    exit(1);
+}
+
+class Condition {
+public:
+    virtual ~Condition() { }
+    virtual bool check() const = 0;
+};
+
+class And : public Condition {
+public:
+    And(NonnullOwnPtr<Condition> lhs, NonnullOwnPtr<Condition> rhs)
+        : m_lhs(move(lhs))
+        , m_rhs(move(rhs))
+    {
+    }
+
+private:
+    virtual bool check() const override
+    {
+        return m_lhs->check() && m_rhs->check();
+    }
+
+    NonnullOwnPtr<Condition> m_lhs;
+    NonnullOwnPtr<Condition> m_rhs;
+};
+
+class Or : public Condition {
+public:
+    Or(NonnullOwnPtr<Condition> lhs, NonnullOwnPtr<Condition> rhs)
+        : m_lhs(move(lhs))
+        , m_rhs(move(rhs))
+    {
+    }
+
+private:
+    virtual bool check() const override
+    {
+        return m_lhs->check() || m_rhs->check();
+    }
+
+    NonnullOwnPtr<Condition> m_lhs;
+    NonnullOwnPtr<Condition> m_rhs;
+};
+
+class Not : public Condition {
+public:
+    Not(NonnullOwnPtr<Condition> cond)
+        : m_cond(move(cond))
+    {
+    }
+
+private:
+    virtual bool check() const override
+    {
+        return !m_cond->check();
+    }
+
+    NonnullOwnPtr<Condition> m_cond;
+};
+
+class Exists : public Condition {
```

Comment:
I'd actually expect this to be a variant of the permission checking condition, with `F_OK`